mypy/nodes.py

```diff
@@ -3017,6 +3017,24 @@ class is generic then it will be a type constructor of higher kind.
         "is_intersection",
     ]
 
+    # Special attributes not collected as protocol members by Python 3.12
+    # See typing._SPECIAL_NAMES
+    EXCLUDED_ATTRIBUTES: Final = frozenset(
+        {
+            "__abstractmethods__",
+            "__annotations__",
+            "__dict__",
+            "__doc__",
+            "__init__",
```

Comment:
CC @AlexWaygood about that

Reply:
I _think_ excluding `__init__` and `__new__` makes sense. As @HexDecimal says, it's what CPython does: https://github.com/python/cpython/blob/4328dc646517f9251bdf6c827014f01c5229e8d9/Lib/typing.py#L1678-L1682

In general, constructors and initialisers aren't thought to be relevant to whether or not a class abides by the Liskov Substitution Principle. In general, there's no guarantee that just because class `X` is a subtype of class `Y`, class `X`'s initialiser will therefore be compatible with class `Y`'s initialiser.

Reply:
Python had a change in 3.11 so that `__init__` methods on protocol base classes are preserved in concrete subclasses of that protocol, but it still doesn't an consider `__init__` method a protocol member for the purposes of `isinstance()` or `issubclass()` checks, I don't think

Reply:
I found an old discussion of this topic at #3823.